File: apps/ruby/design-pattern/08-iterator/test/iterator_test.rb
require_relative "test_helper"
require_relative "../lib/array_iterator"
require_relative "../lib/account"
require_relative "../lib/portfolio"

class IteratorTest < Minitest::Test
  def test_external_iterator
    iterator = ArrayIterator.new([1, 2, 3])
    results = []

    while iterator.has_next?
      results << iterator.next_item
    end

    assert_equal [1, 2, 3], results
  end

  def test_account_is_comparable_by_balance
    ordinary = Account.new("普通預金", 100_000)
    time_deposit = Account.new("定期預金", 500_000)

    assert_operator ordinary, :<, time_deposit
  end

  def test_portfolio_sort
    portfolio = Portfolio.new
    portfolio.add_account(Account.new("定期預金", 500_000))
    portfolio.add_account(Account.new("普通預金", 100_000))

    sorted = portfolio.sort

    assert_equal "普通預金", sorted.first.name
  end
end

File: apps/haskell/design-pattern/08-iterator/test/IteratorTest.hs
module IteratorTest (tests) where

import Iterator
  ( Account (..)
  , AccountType (..)
  , Portfolio (..)
  , accountNames
  , addAccount
  , filterByType
  , mergePortfolios
  , newPortfolio
  , sortByBalance
  , totalBalance
  )
import Test.HUnit

tests :: Test
tests =
  TestLabel "IteratorTest" $
    TestList
      [ TestLabel "testSortByBalance" testSortByBalance
      , TestLabel "testAccountNamesAndFilter" testAccountNamesAndFilter
      , TestLabel "testPortfolioOperations" testPortfolioOperations
      ]

portfolio :: Portfolio
portfolio =
  Portfolio
    [ Account "普通預金" Savings 100000.0
    , Account "当座預金" Checking 50000.0
    , Account "投資口座" Investment 200000.0
    ]

testSortByBalance :: Test
testSortByBalance = TestCase $ do
  let sorted = sortByBalance portfolio
  case (sorted, reverse sorted) of
    (lowest : _, highest : _) -> do
      assertEqual "最小残高" 50000.0 (acctBalance lowest)
      assertEqual "最大残高" 200000.0 (acctBalance highest)
    _ -> assertFailure "口座が並び替えられていません"

testAccountNamesAndFilter :: Test
testAccountNamesAndFilter = TestCase $ do
  assertEqual
    "口座名を列挙できる"
    ["普通預金", "当座預金", "投資口座"]
    (accountNames portfolio)
  assertEqual
    "種別で絞り込める"
    [Account "普通預金" Savings 100000.0]
    (filterByType Savings portfolio)
  assertEqual "合計残高" 350000.0 (totalBalance portfolio)

testPortfolioOperations :: Test
testPortfolioOperations = TestCase $ do
  let portfolioA =
        addAccount (Account "預金A" Savings 100000.0) newPortfolio
      portfolioB =
        addAccount (Account "投資B" Investment 200000.0) newPortfolio
      merged = mergePortfolios portfolioA portfolioB
  assertEqual "追加順で名前を取得できる" ["預金A"] (accountNames portfolioA)
  assertEqual "統合後の合計残高" 300000.0 (totalBalance merged)
